libyul/optimiser/Semantics.h

```diff
@@ -53,11 +53,48 @@ class SideEffectsCollector: public ASTWalker
 		Block const& _ast,
 		std::map<YulString, SideEffects> const* _functionSideEffects = nullptr
 	);
+	SideEffectsCollector(
+		Dialect const& _dialect,
+		ForLoop const& _ast,
+		std::map<YulString, SideEffects> const* _functionSideEffects = nullptr
+	);
 
 	using ASTWalker::operator();
 	void operator()(FunctionCall const& _functionCall) override;
 
 	bool movable() const { return cannotLoop() && m_sideEffects.movable; }
+	bool movableRelativeTo(SideEffects const& _other)
+	{
+		if (!m_sideEffects.cannotLoop)
+			return false;
+
+		if (m_sideEffects.movable)
+			return true;
+
```

Comment:
I think it should be fixable by just checking `sideEffectFree[IfNoMSize]` and returning false if this return false.

Reply:
GAS and PC are `sideEffectFree.` So that won't help. It would prevent moving of INVALID and STOP, though. [Here](https://gist.github.com/hrkrshnn/4561d363aa7e12ddea1be3a6a0e7c5e3#file-onlysideeffects-org) is a list of instructions with `stideEffectFree = false`.

Reply:
I think we need a new flag, say `otherImmovables`, that handles gas, pc, stop and invalid.

Reply:
`PC` is totally incompatible with the optimzier, so I think it is fine to set its 'sideEffectFree' flag to false.

Why shouldn't `gas` have `sideEffectFree` set to false?

Reply:
The list you linked is based on `Instructions.cpp`, right? We should refine that list for the yul optimizer, it is too imprecise now that we have the detailed effects on memory and storage.

Reply:
`sideEffectFree` seems to be only used for `UnusedPruner `, and it makes sense for `gas` to be `sideEffectFree.` I think a new parameter in the `SideEffects` struct is the only way around.

The list is based only on `Instructions.cpp.` I can make a new list collecting the `SideEffects` struct for instructions. It is currently not obvious from the code what instructions can get moved out of the loop, and the list can be quite useful.

Reply:
Yeah, you are probably right. The thing is that `gas` is not movable but also neither reads nor writes the state. Maybe we should also rename `sideEffectFree` to `canBeRemoved` - because that's what it is used for and that's also what is written in its description.